services/httpd/msgp_emitter.go

```diff
@@ -0,0 +1,160 @@
+package httpd
+
+import (
+	"io"
+	"net/http"
+
+	"github.com/influxdata/influxdb/influxql"
+	"github.com/tinylib/msgp/msgp"
+)
+
+type messagePackEncoder struct {
+	Epoch     string
+	ChunkSize int
+}
+
+func (e *messagePackEncoder) ContentType() string {
+	return "application/x-msgpack"
+}
+
+func (e *messagePackEncoder) Encode(w io.Writer, header ResponseHeader, results <-chan *influxql.ResultSet) {
+	var convertToEpoch func(row *influxql.Row)
+	if e.Epoch != "" {
+		convertToEpoch = epochConverter(e.Epoch)
+	}
+	values := make([][]interface{}, 0, e.ChunkSize)
+
+	enc := msgp.NewWriter(w)
+	enc.WriteMapHeader(1)
+	enc.WriteString("results")
+	enc.WriteInt(header.Results)
+
+	enc.Flush()
+	if w, ok := w.(http.Flusher); ok {
+		w.Flush()
+	}
+
+	for result := range results {
+		var messages []Message
+		if len(result.Messages) > 0 {
+			messages = make([]Message, len(result.Messages))
+			for i, m := range result.Messages {
+				messages[i].Level = m.Level
+				messages[i].Text = m.Text
+			}
+		}
+		header := ResultHeader{
+			ID:       result.ID,
+			Messages: messages,
+		}
+		if result.Err != nil {
+			err := result.Err.Error()
+			header.Error = &err
+		}
+		header.EncodeMsg(enc)
+		if result.Err != nil {
+			enc.Flush()
+			if w, ok := w.(http.Flusher); ok {
+				w.Flush()
+			}
+			continue
+		}
+
+		for series := range result.SeriesCh() {
+			enc.WriteArrayHeader(2)
+			enc.WriteInt(1)
+			enc.WriteBool(true)
+
+			if series.Err != nil {
+				err := series.Err.Error()
+				header := SeriesError{Error: err}
+				header.EncodeMsg(enc)
+
+				enc.Flush()
+				if w, ok := w.(http.Flusher); ok {
+					w.Flush()
+				}
+				continue
+			}
+
+			columns := make([]Column, len(series.Columns))
+			for i, col := range series.Columns {
+				columns[i] = Column{Name: col.Name, Type: col.Type.String()}
+			}
+			header := SeriesHeader{
+				Name:    &series.Name,
+				Tags:    series.Tags.KeyValues(),
+				Columns: columns,
+			}
+			header.EncodeMsg(enc)
+
+			for row := range series.RowCh() {
+				if row.Err != nil {
+					enc.WriteArrayHeader(2)
+					enc.WriteInt(len(values) + 1)
+					enc.WriteBool(false)
+					for _, v := range values {
+						val := Row{Value: v}
+						val.EncodeMsg(enc)
+					}
+					values = values[:0]
+
+					err := RowError{Error: row.Err.Error()}
+					err.EncodeMsg(enc)
+					continue
+				}
+
+				if convertToEpoch != nil {
+					convertToEpoch(&row)
+				}
+
+				if len(values) == cap(values) {
+					enc.WriteArrayHeader(2)
+					enc.WriteInt(len(values))
```

Comment:
If you can give me an idea what you mean by true streaming, I'm all for that. The old response structure had a problem that it wasn't true streaming and the results were embedded in a weird way that made reading from it as a client impossible.

The reason why this encodes each row is so that it can read one row at a time. If you encode as an array, you have to decode the entire array rather than just the next row. Then you have to buffer the full results from the chunk. While this method requires maintaining state for how much you are going to read, that's just an integer and a boolean rather than an array. And you would have to still maintain state for the boolean (do I need to read more when I finish reading the array?) and for the array itself. So both maintain roughly the same state.

Now reading an array of rows as a chunk may perform better. If it does, we should consider the pros and cons. I haven't performed any testing of that.